### PR TITLE
feat: Sync timeouts for applications (#6055)

### PR DESCRIPTION
**Pull request by andrii-korotkov-verkada**
Helps with #6055

Introduces a controller-level configuration for terminating sync after timeout.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by rumstead** [comment]
Nice! Do you think we can add a few e2e tests?

**Comment by andrii-korotkov-verkada** [comment]
I have `TestSyncTimeout`, but can also add end-to-end if necessary. Though I'm not sure they'd add more value, since they'd be replicating the logic from `TestSyncTimeout`. What do you think?

**Comment by rumstead** [comment]
> I have `TestSyncTimeout`, but can also add end-to-end if necessary. Though I'm not sure they'd add more value, since they'd be replicating the logic from `TestSyncTimeout`. What do you think?

I was thinking that deploying an application that will never leave progressing and being able to make sure it the sync status fails would be a good one to have. If you think the unit tests are solid, rock on!

**Review by ishitasequeira — COMMENTED** [review]
Overall PR looks good. Although, I have 1 minor comment.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 1]
```diff
@@ -248,6 +250,7 @@ func NewCommand() *cobra.Command {
 	command.Flags().IntVar(&selfHealBackoffTimeoutSeconds, "self-heal-backoff-timeout-seconds", env.ParseNumFromEnv("ARGOCD_APPLICATION_CONTROLLER_SELF_HEAL_BACKOFF_TIMEOUT_SECONDS", 2, 0, math.MaxInt32), "Specifies initial timeout of exponential backoff between self heal attempts")
 	command.Flags().IntVar(&selfHealBackoffFactor, "self-heal-backoff-factor", env.ParseNumFromEnv("ARGOCD_APPLICATION_CONTROLLER_SELF_HEAL_BACKOFF_FACTOR", 3, 0, math.MaxInt32), "Specifies factor of exponential timeout between application self heal attempts")
 	command.Flags().IntVar(&selfHealBackoffCapSeconds, "self-heal-backoff-cap-seconds", env.ParseNumFromEnv("ARGOCD_APPLICATION_CONTROLLER_SELF_HEAL_BACKOFF_CAP_SECONDS", 300, 0, math.MaxInt32), "Specifies max timeout of exponential backoff between application self heal attempts")
+	command.Flags().IntVar(&syncTimeout, "sync-timeout", env.ParseNumFromEnv("ARGOCD_APPLICATION_CONTROLLER_SYNCTIMEOUT", 0, 0, math.MaxInt32), "Specifies the timeout after which a sync would be terminated. 0 means no timeout (default 0).")
```
[the remark this answers is not in the document]
Comment (andrii-korotkov-verkada) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go]: Good catch! Fixed.

**Review by ishitasequeira — APPROVED** [review]
Changes look good!!

**Comment by jhansireddy01** [comment]
After upgrading to the rc 5 version of 2.14.0, we tested this feature by adding the following configuration to argocd-cmd-params-cm:

**controller.sync.timeout.seconds: '30'**

However, even after reproducing the never-ending syncing state, it does not terminate after the specified timeout.
Could you advise if there is any misconfiguration on our end?

**Comment by andrii-korotkov-verkada** [comment]
I suspect 30 may be too small. Can you try with a value which would be more practical like 900-1800?

**Comment by jhansireddy01** [comment]
> I suspect 30 may be too small. Can you try with a value which would be more practical like 900-1800?

Tried introducing a timeout of 30 minutes, it wont work 
Screenshot from 2025-01-22 12-24-58
Screenshot from 2025-01-22 12-27-54
image

**Comment by andrii-korotkov-verkada** [comment]
Ah, I think maybe it's because it keep retrying and counts each attempt as a separate sync? I can confirm that I observed this feature working in at least one case in my company. Can you try disabling infinite retries and see what happens?

**Comment by jhansireddy01** [comment]
> Ah, I think maybe it's because it keep retrying and counts each attempt as a separate sync? I can confirm that I observed this feature working in at least one case in my company. Can you try disabling infinite retries and see what happens?

can i get the scenario where you tested this on, I will try to reproduce it ?

**Comment by andrii-korotkov-verkada** [comment]
> can i get the scenario where you tested this on, I will try to reproduce it ?

I don't have a repro :(

**Comment by jhansireddy01** [comment]
> > can i get the scenario where you tested this on, I will try to reproduce it ?
> 
> I don't have a repro :(

managed to reproduce a never ending sync state and observed the sync failing without any manual intervention
image
Tho the sync went on for 31 minutes when i set a 30 minute timeout, Is this normal?

**Comment by jhansireddy01** [comment]
image
this is one more scenario where i set the sync timeout  for 15 minutes, looks like its working :)
buts somehow if there are retries looks like it wont work if  im not wrong

**Comment by andrii-korotkov-verkada** [comment]
> Tho the sync went on for 31 minutes when i set a 30 minute timeout, Is this normal?

I'm not sure, check out various status fields and see if the start timestamp is somewhat different than when the sync has started. I've noticed the sync initially is "waiting to start", so maybe that's not counted towards timeout.

**Pull request merged**